README.txt
offline_proxy
=============

The purpose of this package is to provide a simple proxy which stores
the response of successful requests locally on the disk.

The main use case is provide a solid web service for TV screens showing
informations from a website. In a environment with a really instable Internet connection, or WIFI connection.
Thus this package keeps everything a simple as possible

Limitations
-----------

- Only GET requests are handled.
- Conent-Encoding header not supported.
- No performance improvement at all - in fact it adds a lot of overhead.


Development
-----------

Run single worker locally for debugging purposes.

```
python offline_proxy/server.py --backend=https://some.site --port=8080 --storage=/path/to/cache
```


Install and run tests:

```
pip install .[tests]
pytest
```


Deployment
----------

```
pip install offline_proxy
```

See https://docs.aiohttp.org/en/stable/deployment.html for detailed informations

Here is an example using gunicorn, which is shipped with this packages too:

```
gunicorn offline_proxy.server:app --bind localhost:8080 --worker-class aiohttp.GunicornWebWorker --workers 4 --env PROXY_BACKEND="https://some.site" --env PROXY_CACHE_LOCATION=/path/to/cache
```

The service requires to configurations in your environment:

- PROXY_BACKEND
- PROXY_CACHE_LOCATION

They need to be present, otherwise the server doesn't start.



CHANGES.txt
Changelog
=========

1.0.0.dev
---------

- Init Release with basic functionality [mathias]


local_proxy/tests/test_proxy.py
from aioresponses import aioresponses
from local_proxy.cache import ProxyCache
from local_proxy.server import app
import aiohttp
import pytest
import tempfile
import shutil


TEST_BACKEND = 'http://test.com'


@pytest.fixture()
async def app_server(loop, aiohttp_server, tmpdir_factory):
    cache_dir = tmpdir_factory.mktemp('cache')
    proxy_app = await app(backend=TEST_BACKEND, location=cache_dir)
    server = await aiohttp_server(proxy_app)
    server.url = f'http://localhost:{server.port}'
    server.cache_dir = cache_dir
    yield server
    shutil.rmtree(cache_dir)


async def client(endpoint):
    async with aiohttp.request('GET', endpoint) as response:
        body = await response.content.read()
    return response, body


class TestProxy:

    async def test_app_needs_a_backend(self):
        with pytest.raises(Exception):
            await app()

    async def test_wildcard_enpoint(self, app_server):

        with aioresponses(passthrough=[app_server.url]) as mock:
            mock.get(rf'{TEST_BACKEND}.*', status='200', body='OK')

            endpoints = [
                '/test',
                '/test/',
                '/test/long/path',
                '/test?param=something',
                '/something?param=test&param2=test2'
            ]
            for endpoint in endpoints:
                response, body = await client(app_server.url + endpoint)
                assert endpoint == response.url.raw_path_qs

    async def test_return_fallback_response_on_backend_error(self, app_server):

        with aioresponses(passthrough=[app_server.url]) as mock:
            mock.get(TEST_BACKEND + '/error', status='500')

            response, body = await client(app_server.url + '/error')

            assert 404 == response.status


    async def test_requests_are_getting_cached(self, app_server):

        with aioresponses(passthrough=[app_server.url]) as mock:
            mock.get(TEST_BACKEND + '/site', status='200', body='OK')

            response, body = await client(app_server.url + '/site')

            cache = ProxyCache(app_server.cache_dir)
            assert 2 == len(cache.cache)
            assert body == cache.get_cached_content_by_url(TEST_BACKEND + '/site')


    async def test_request_returns_cached_content_on_server_error(self, app_server):

        with aioresponses(passthrough=[app_server.url]) as mock:
            mock.get(TEST_BACKEND + '/site', status='200', body='Cached')
            response, body = await client(app_server.url + '/site')

        with aioresponses(passthrough=[app_server.url]) as mock:
            mock.get(TEST_BACKEND + '/site', status='500')
            response, body = await client(app_server.url + '/site')

            assert b'Cached' == body

    async def test_request_returns_not_found_if_not_in_cache(self, app_server):

        with aioresponses(passthrough=[app_server.url]) as mock:
            mock.get(TEST_BACKEND + '/site', status='200', body='Cached')
            response, body = await client(app_server.url + '/site')


        with aioresponses(passthrough=[app_server.url]) as mock:
            mock.get(TEST_BACKEND + '/anothersite', status='500')
            response, body = await client(app_server.url + '/anothersite')

            assert 404 == response.status


offline_proxy/tests/test_proxy.py
from aioresponses import aioresponses
from offline_proxy.cache import ProxyCache
from offline_proxy.server import app
import aiohttp
import pytest
import shutil


TEST_BACKEND = 'http://test.com'


@pytest.fixture()
async def app_server(loop, aiohttp_server, tmpdir_factory):
    cache_dir = tmpdir_factory.mktemp('cache')
    proxy_app = await app(backend=TEST_BACKEND, location=cache_dir)
    server = await aiohttp_server(proxy_app)
    server.url = f'http://localhost:{server.port}'
    server.cache_dir = cache_dir
    yield server
    shutil.rmtree(cache_dir)


async def client(endpoint):
    async with aiohttp.request('GET', endpoint) as response:
        body = await response.content.read()
    return response, body


class TestProxy:

    async def test_app_needs_a_backend(self):
        with pytest.raises(Exception):
            await app()

    async def test_wildcard_enpoint(self, app_server):

        with aioresponses(passthrough=[app_server.url]) as mock:
            mock.get(rf'{TEST_BACKEND}.*', status='200', body='OK')

            endpoints = [
                '/test',
                '/test/',
                '/test/long/path',
                '/test?param=something',
                '/something?param=test&param2=test2'
            ]
            for endpoint in endpoints:
                response, body = await client(app_server.url + endpoint)
                assert endpoint == response.url.raw_path_qs

    async def test_return_fallback_response_on_backend_error(self, app_server):

        with aioresponses(passthrough=[app_server.url]) as mock:
            mock.get(TEST_BACKEND + '/error', status='500')

            response, body = await client(app_server.url + '/error')

            assert 404 == response.status

    async def test_requests_are_getting_cached(self, app_server):

        with aioresponses(passthrough=[app_server.url]) as mock:
            mock.get(TEST_BACKEND + '/site', status='200', body='OK')

            response, body = await client(app_server.url + '/site')

            cache = ProxyCache(app_server.cache_dir)
            assert 2 == len(cache.cache)
            assert body == cache.get_cached_content_by_url(TEST_BACKEND + '/site')

    async def test_request_returns_cached_content_on_server_error(self, app_server):

        with aioresponses(passthrough=[app_server.url]) as mock:
            mock.get(TEST_BACKEND + '/site', status='200', body='Cached')
            response, body = await client(app_server.url + '/site')

        with aioresponses(passthrough=[app_server.url]) as mock:
            mock.get(TEST_BACKEND + '/site', status='500')
            response, body = await client(app_server.url + '/site')

            assert b'Cached' == body

    async def test_request_returns_not_found_if_not_in_cache(self, app_server):

        with aioresponses(passthrough=[app_server.url]) as mock:
            mock.get(TEST_BACKEND + '/site', status='200', body='Cached')
            response, body = await client(app_server.url + '/site')

        with aioresponses(passthrough=[app_server.url]) as mock:
            mock.get(TEST_BACKEND + '/anothersite', status='500')
            response, body = await client(app_server.url + '/anothersite')

            assert 404 == response.status


offline_proxy/cache.py
from diskcache import Cache
import hashlib
import pickle


class ProxyCache:
    cache = None

    def __init__(self, location, size_limit=2 ** 29):
        """
        A location needs to be provided.
        Default size is 500MB
        """

        self.cache = Cache(location, expire=None, size_limit=size_limit)

    @staticmethod
    def get_key(url):
        return hashlib.md5(url.encode()).hexdigest()

    def add_respone_to_cache(self, url, headers, body):
        """
        Adds a new item to the cache using the url.
        """
        key = self.get_key(url)
        self.cache[key] = body
        self.cache[key + '__header'] = pickle.dumps(headers)

    def get_cached_content_by_url(self, url):
        return self.cache[self.get_key(url)]

    def get_cached_header_by_url(self, url):
        return pickle.loads(self.cache[self.get_key(url) + '__header'])

    def __contains__(self, url):
        return self.get_key(url) in self.cache


offline_proxy/server.py
from aiohttp.client_exceptions import ClientConnectorError
from aiohttp.client_exceptions import ServerTimeoutError
from offline_proxy.cache import ProxyCache
from urllib.parse import urljoin
import aiohttp.web
import argparse
import logging
import os
import socket


routes = aiohttp.web.RouteTableDef()
logger = logging.getLogger("local proxy")


parser = argparse.ArgumentParser(description='Simple local proxy configuration')
parser.add_argument('--backend', required=True)
parser.add_argument('--location', required=True)
parser.add_argument('--port', required=True)


class ServerError(Exception):
    pass


class LocalProxy:

    backend = None
    cache = None

    def __init__(self, backend, location):
        self.backend = backend
        self.cache = ProxyCache(location)

    async def endpoint(self, request):
        target = urljoin(self.backend, request.raw_path)
        logger.info(f'Distpach to: {target}')

        try:
            timeout = aiohttp.ClientTimeout(total=60)
            async with aiohttp.request('GET', target, timeout=timeout) as response:
                body = await response.content.read()

            if str(response.status).startswith('5'):
                raise ServerError()

            headers = dict(response.headers)

            # Prepare headers for response
            headers.pop('Content-Encoding', None)
            headers['Content-Length'] = str(len(body))

            self.cache.add_respone_to_cache(target, headers, body)
            return aiohttp.web.Response(
                body=body,
                status=response.status,
                headers=headers
            )

        except (socket.gaierror, ClientConnectorError, ServerTimeoutError, ServerError):

            if target in self.cache:
                return aiohttp.web.Response(
                    body=self.cache.get_cached_content_by_url(target),
                    status=200,
                    headers=self.cache.get_cached_header_by_url(target)
                )
            else:
                return aiohttp.web.Response(
                    status=404,
                    content_type='text/plain'
                )


async def app(backend=None, location=None):

    env_backend = os.environ.get('PROXY_BACKEND')
    env_location = os.environ.get('PROXY_CACHE_LOCATION')

    if backend is None and env_backend is None:
        raise Exception('PROXY_BACKEND env var is required')
    elif backend is None and env_backend:
        backend = env_backend

    if location is None and env_location is None:
        raise Exception('PROXY_CACHE_LOCATION env var is required')
    elif location is None and env_location:
        location = env_location

    proxy = LocalProxy(backend, location)
    proxy_app = aiohttp.web.Application()
    proxy_app.add_routes(
        [
            aiohttp.web.get('/{tail:.*}', proxy.endpoint)
        ]
    )
    return proxy_app


if __name__ == "__main__":
    args = parser.parse_args()
    aiohttp.web.run_app(app(args.backend, args.location), port=args.port)
